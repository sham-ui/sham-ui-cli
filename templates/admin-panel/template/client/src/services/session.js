import { DI, inject } from 'sham-ui';

export default class Session {
    @inject router;
    @inject store;
    @inject( 'session:storage' ) data;

    constructor() {
        DI.bind( 'session', this );
    }

    login( email, password ) {
        return this.store.login( { email, password } ).then( ( { Email, Name } ) => {
            this.data.email = Email;
            this.data.name = Name;
        } ).then(
            ::this.resetSessionValidation
        ).then(
            () => this.validateSessionPromise
        );
    }

    get validateSessionPromise() {
        if ( undefined === this._validateSessionPromise ) {
            this._validateSessionPromise = this.validateSession();
        }
        return this._validateSessionPromise;
    }

    resetSessionValidation() {
        this._validateSessionPromise = undefined;
    }

    validateSession() {
        this.data.sessionValidated = false;
        return this.store.validSession().then(
            ( { Email, Name } ) => {
                this.data.sessionValidated = true;
                this.data.isAuthenticated = true;
                this.data.email = Email;
                this.data.name = Name;

                // Manual run sync for guaranteed update Layout
                // component before promise resolved
                this.data.sync();
                return true;
            },
            () => {
                this.data.sessionValidated = true;
                this.data.isAuthenticated = false;
                this._goToLoginPage();
                return false;
            }
        );
    }

    logout() {
        this.store.logout().then(
            () => {
                this.resetSessionValidation();
                this._goToLoginPage();
            }
        );
    }

    _isRestrictedPage() {
        const page = this.router.lastRouteResolved();
        return undefined !== page && ![
            'signup',
            'login'
        ].includes( page.name );
    }

    _goToLoginPage() {
        this.data.isAuthenticated = false;
        this.data.email = '';
        this.data.name = '';
        if ( this._isRestrictedPage() ) {
            requestAnimationFrame(
                () => this.router.navigate(
                    this.router.generate( 'login', {} )
                )
            );
        }
    }
}
