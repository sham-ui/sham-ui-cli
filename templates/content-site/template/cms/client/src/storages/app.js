import createStorage from 'sham-ui-data-storage';

export const { storage, useStorage } = createStorage( {
    routerResolved: false,
    tokenLoaded: false
}, {
    DI: 'app:storage',
    sync: true
} );
