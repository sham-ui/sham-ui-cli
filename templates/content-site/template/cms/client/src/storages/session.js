import createStorage from 'sham-ui-data-storage';

export const { storage, useStorage } = createStorage( {
    name: '',
    email: '',
    sessionValidated: false,
    isAuthenticated: false,
    isSuperuser: false
}, {
    DI: 'session:storage'
} );
