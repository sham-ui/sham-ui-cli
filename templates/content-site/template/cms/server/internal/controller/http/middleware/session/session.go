package session

import (
	"cms/internal/controller/http/request"
	"cms/internal/model"
	"cms/pkg/logger"
	"net/http"

	"go.opentelemetry.io/otel/attribute"
	"go.opentelemetry.io/otel/trace"
)

type Service interface {
	Get(r *http.Request) (*model.Session, error)
}

type session struct {
	srv Service
}

func (sess *session) Middleware(next http.Handler) http.Handler {
	return http.HandlerFunc(
		func(rw http.ResponseWriter, r *http.Request) {
			memberSess, err := sess.srv.Get(r)
			if err == nil {
				ctx := r.Context()

				// Save session to trace context
				span := trace.SpanFromContext(ctx)
				if span.SpanContext().IsValid() {
					span.SetAttributes(
						attribute.String("member_id", string(memberSess.MemberID)),
					)
				}

				// Save session to logger context
				log := logger.Load(ctx).WithValues(
					"member_id", string(memberSess.MemberID),
				)
				ctx = logger.Save(r.Context(), log)

				// Save session to request context
				ctx = request.SaveSessionToContext(ctx, memberSess)
				r = r.WithContext(ctx)
			}
			next.ServeHTTP(rw, r)
		},
	)
}

func New(srv Service) *session {
	return &session{
		srv: srv,
	}
}
