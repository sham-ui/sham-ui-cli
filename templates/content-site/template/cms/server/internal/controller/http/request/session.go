package request

import (
	"cms/internal/model"
	"context"
)

type contextKeySession struct{}

func SessionFromContext(ctx context.Context) (*model.Session, bool) {
	sess, ok := ctx.Value(contextKeySession{}).(*model.Session)
	return sess, ok
}

func SaveSessionToContext(ctx context.Context, session *model.Session) context.Context {
	return context.WithValue(ctx, contextKeySession{}, session)
}
