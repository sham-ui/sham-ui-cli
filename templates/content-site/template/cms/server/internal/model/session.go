package model

import "errors"

type Session struct {
	MemberID    MemberID
	Email       string
	Name        string
	IsSuperuser bool
}

var (
	ErrSessionNotAuthenticated = errors.New("session not authenticated")
	ErrSessionNotSuperuser     = errors.New("session not superuser")
)
