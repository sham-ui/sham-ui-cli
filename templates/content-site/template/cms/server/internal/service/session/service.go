package session

import (
	"cms/internal/model"
	"cms/pkg/postgres"
	"cms/pkg/tracing"
	"context"
	"fmt"
	"net/http"
	"time"

	"github.com/antonlindstrom/pgstore"
)

const (
	scopeName = "service.session"

	sessionName = "cms-session"

	sessionValueAuthenticated = "authenticated"
	sessionValueMemberID      = "id"
	sessionValueEmail         = "email"
	sessionValueName          = "name"
	sessionValueIsSuperuser   = "is_superuser"
)

type service struct {
	store *pgstore.PGStore

	quit chan<- struct{}
	done <-chan struct{}
}

func (s *service) String() string {
	return scopeName
}

func (s *service) GracefulShutdown(_ context.Context) error {
	s.store.StopCleanup(s.quit, s.done)
	return nil
}

func (s *service) Get(r *http.Request) (*model.Session, error) {
	const op = "Get"

	_, span := tracing.StartSpan(r.Context(), scopeName, op)
	defer span.End()

	raw, err := s.store.Get(r, sessionName)
	if err != nil {
		return nil, fmt.Errorf("get session from store: %w", err)
	}
	if auth, err := getValue[bool](raw, sessionValueAuthenticated); err != nil || !auth {
		return nil, model.ErrSessionNotAuthenticated
	}
	session := &model.Session{} //nolint:exhaustruct
	memberID, err := getValue[string](raw, sessionValueMemberID)
	if err != nil {
		return nil, fmt.Errorf("get member id: %w", err)
	}
	session.MemberID = model.MemberID(memberID)
	if session.Name, err = getValue[string](raw, sessionValueName); err != nil {
		return nil, fmt.Errorf("get name: %w", err)
	}
	if session.Email, err = getValue[string](raw, sessionValueEmail); err != nil {
		return nil, fmt.Errorf("get email: %w", err)
	}
	if session.IsSuperuser, err = getValue[bool](raw, sessionValueIsSuperuser); err != nil {
		return nil, fmt.Errorf("get is_superuser: %w", err)
	}
	return session, nil
}

func (s *service) Create(rw http.ResponseWriter, r *http.Request, member *model.Member) error {
	const op = "Create"

	_, span := tracing.StartSpan(r.Context(), scopeName, op)
	defer span.End()

	raw, err := s.store.New(r, sessionName)
	if err != nil {
		return fmt.Errorf("can't create new session: %w", err)
	}
	raw.Values[sessionValueAuthenticated] = true
	raw.Values[sessionValueMemberID] = string(member.ID)
	raw.Values[sessionValueName] = member.Name
	raw.Values[sessionValueEmail] = member.Email
	raw.Values[sessionValueIsSuperuser] = member.IsSuperuser
	if err := raw.Save(r, rw); err != nil {
		return fmt.Errorf("can't save session: %w", err)
	}
	return nil
}

func (s *service) Delete(rw http.ResponseWriter, r *http.Request) error {
	const op = "Delete"

	_, span := tracing.StartSpan(r.Context(), scopeName, op)
	defer span.End()

	raw, err := s.store.Get(r, sessionName)
	if err != nil {
		return fmt.Errorf("get session from store: %w", err)
	}
	raw.Values[sessionValueAuthenticated] = false
	raw.Options.MaxAge = -1
	if err := raw.Save(r, rw); err != nil {
		return fmt.Errorf("can't save session: %w", err)
	}
	return nil
}

func (s *service) UpdateEmail(rw http.ResponseWriter, r *http.Request, email string) error {
	const op = "UpdateEmail"
	_, span := tracing.StartSpan(r.Context(), scopeName, op)
	defer span.End()
	if err := s.updateValue(rw, r, sessionValueEmail, email); err != nil {
		return fmt.Errorf("update email: %w", err)
	}
	return nil
}

func (s *service) UpdateName(rw http.ResponseWriter, r *http.Request, name string) error {
	const op = "UpdateName"
	_, span := tracing.StartSpan(r.Context(), scopeName, op)
	defer span.End()
	if err := s.updateValue(rw, r, sessionValueName, name); err != nil {
		return fmt.Errorf("update name: %w", err)
	}
	return nil
}

func (s *service) updateValue(rw http.ResponseWriter, r *http.Request, field string, value any) error {
	raw, err := s.store.Get(r, sessionName)
	if err != nil {
		return fmt.Errorf("get session from store: %w", err)
	}
	raw.Values[field] = value
	if err := raw.Save(r, rw); err != nil {
		return fmt.Errorf("can't save session: %w", err)
	}
	return nil
}

func New(
	db *postgres.Database,
	secret string,
	domain string,
	maxAge time.Duration,
) (*service, error) {
	store, err := pgstore.NewPGStoreFromPool(db.Postgres(), []byte(secret))
	if err != nil {
		return nil, fmt.Errorf("can't create pgstore: %w", err)
	}
	store.MaxAge(int(maxAge.Seconds()))
	store.Options.Domain = domain
	quit, done := store.Cleanup(maxAge)
	return &service{
		store: store,
		quit:  quit,
		done:  done,
	}, nil
}
