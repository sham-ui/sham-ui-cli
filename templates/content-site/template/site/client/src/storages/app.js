import createStorage from 'sham-ui-data-storage';

export const { storage, useStorage } = createStorage( {
    routerResolved: false,
    darkThemeEnabled: false
}, {
    DI: 'app:storage',
    sync: true
} );
