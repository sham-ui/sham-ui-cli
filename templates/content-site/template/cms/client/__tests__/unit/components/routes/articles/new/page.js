import { createDI } from 'sham-ui';
import * as directives from 'sham-ui-directives';
import RoutesArticlesNewPage  from '../../../../../../src/components/routes/articles/new/page.sfc';
import renderer from 'sham-ui-test-helpers';
import { storage } from '../../../../../../src/storages/session';

it( 'renders correctly', () => {
    const DI = createDI();
    DI.bind( 'title', {
        change() {}
    } );
    storage( DI ).sessionValidated = true;

    const getMock = jest.fn();
    DI.bind( 'store', {
        articleCategories: getMock.mockReturnValueOnce(
            Promise.resolve( {
                categories: [
                    { 'id': '4', 'name': 'Hello world!', 'slug': 'hello-world' },
                    { 'id': '6', 'name': 'Название категории', 'slug': 'nazvanie-kategorii' }
                ],
                meta: {}
            } )
        ),
        articleTags: getMock.mockReturnValueOnce(
            Promise.resolve( {
                tags: [
                    { 'id': '4', 'name': 'Hello world!', 'slug': 'hello-world' },
                    { 'id': '6', 'name': 'Название тега', 'slug': 'nazvanie-tega' }
                ],
                meta: {}
            } )
        )
    } );

    const meta = renderer( RoutesArticlesNewPage, {
        DI,
        directives
    } );
    expect( meta.toJSON() ).toMatchSnapshot();
} );
