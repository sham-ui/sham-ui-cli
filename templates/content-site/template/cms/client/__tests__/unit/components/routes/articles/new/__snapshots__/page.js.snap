// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
Object {
  "Options": Object {
    "errors": Array [],
    "pageTitle": "Create article",
  },
  "html": "
<h4 class=\\"page-header\\"> Create article
  <!--0-->
</h4>
<div class=\\"page-body\\">
  <div class=\\"articles-page\\">
    <!--0-->
    <!--0-->
    <div class=\\"panel\\">
      <!--0-->
      <!--0-->
      <!--0-->
    </div>
  </div>
  <!--1-->
</div>
<!--0-->
<!--0-->
",
}
`;
